zeroconf/__init__.py

```diff
@@ -2766,69 +2767,77 @@ def handle_query(self, msg: DNSIncoming, addr: Optional[str], port: int) -> None
                 out.add_question(question)
 
         for question in msg.questions:
-            if question.type == _TYPE_PTR:
-                if question.name == "_services._dns-sd._udp.local.":
-                    for stype in self.servicetypes.keys():
-                        if out is None:
-                            out = DNSOutgoing(_FLAGS_QR_RESPONSE | _FLAGS_AA)
-                        out.add_answer(
-                            msg,
-                            DNSPointer(
-                                "_services._dns-sd._udp.local.", _TYPE_PTR, _CLASS_IN, _DNS_OTHER_TTL, stype
-                            ),
-                        )
-                for service in self.services.values():
-                    if question.name == service.type:
-                        if out is None:
-                            out = DNSOutgoing(_FLAGS_QR_RESPONSE | _FLAGS_AA)
-                        out.add_answer(
-                            msg,
-                            DNSPointer(service.type, _TYPE_PTR, _CLASS_IN, service.other_ttl, service.name),
-                        )
-
-                        # Add recommended additional answers according to
-                        # https://tools.ietf.org/html/rfc6763#section-12.1.
-                        out.add_additional_answer(
-                            DNSService(
-                                service.name,
-                                _TYPE_SRV,
-                                _CLASS_IN | _CLASS_UNIQUE,
-                                service.host_ttl,
-                                service.priority,
-                                service.weight,
-                                cast(int, service.port),
-                                service.server,
+            try:
+                if question.type == _TYPE_PTR:
+                    if question.name == "_services._dns-sd._udp.local.":
+                        for stype in self.servicetypes.keys():
+                            if out is None:
+                                out = DNSOutgoing(_FLAGS_QR_RESPONSE | _FLAGS_AA)
+                            out.add_answer(
+                                msg,
+                                DNSPointer(
+                                    "_services._dns-sd._udp.local.",
+                                    _TYPE_PTR,
+                                    _CLASS_IN,
+                                    _DNS_OTHER_TTL,
+                                    stype,
+                                ),
                             )
-                        )
-                        out.add_additional_answer(
-                            DNSText(
-                                service.name,
-                                _TYPE_TXT,
-                                _CLASS_IN | _CLASS_UNIQUE,
-                                service.other_ttl,
-                                service.text,
+                    for service in self.services.values():
```

Comment:
I could modify `self.servicetypes` under the service lock to make it a `dict` of types with the values as a list of `ServiceInfo` so we can do 

`self.servicetypes.get(question.name)` and then copy the list so we only copy on a hit.